micrometer-core/src/main/java/io/micrometer/core/instrument/DynamicValueTag.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2020 VMware, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument;
+
+import io.micrometer.core.lang.Nullable;
+
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Dynamic value {@link Tag}.
+ *
+ * @author Simon Scholz
+ */
+final class DynamicValueTag implements Tag {
+
+    private final String key;
+    private final Supplier<String> valueSupplier;
+
+    DynamicValueTag(String key, Supplier<String> valueSupplier) {
+        this.key = requireNonNull(key, "The key must not be null");
+        this.valueSupplier = requireNonNull(valueSupplier, "The valueSupplier must not be null");;
+    }
+
+    @Override
+    public String getKey() {
+        return key;
+    }
+
+    @Override
+    public String getValue() {
+        String value = valueSupplier.get();
+        return value != null ? value : "";
+    }
+
+    @Override
+    public boolean equals(@Nullable Object o) {
```

Comment:
`MeterRegistry` stores all meters in a map, where `Meter.Id` is used as key. But this means that `Meter.Id.hashcode()` or `Meter.Id.equals()` will be called and must not change over time. But because
`Meter.Id.hashcode()` calls `Tags.hashcode()` which calls `DynamicValueTag.hashcode()`, your implementation must only take the key into account to show the required stable bahavior. However I'm not completely sure if its an issue to not take the value into account in equality checks. Maybe thats the issue jkschneider was talking about.

Reply:
Let's see what @jkschneider or maybe @izeye think about the `equals` check.